KIm documentation for NS2/CN (#970)
kyma-project/kyma-infrastructure-manager: docs/user/README.md

# Kyma Infrastructure Manager

## Introduction

Kyma Infrastructure Manager (KIM) manages the lifecycle of SAP BTP, Kyma runtime by using SAP Gardener to create and manage Kubernetes cluster-related infrastructure components.

KIM translates Kubernetes cluster-related requirements for Kyma runtime into Gardener's Shoot definitions. It creates, updates, and deletes the Shoot definition. While Gardener reconciles a Kubernetes cluster, KIM monitors the progress, reacting to failure cases and ensuring the cluster is prepared for running the Kyma software.

> ## Constraints

Author: ```suggestion
## Stakeholder Expectations
```

KIM must be implemented as a Kubernetes Operator. It provides a Custom Resource Definition (CRD) that exposes all configurable options of a Kubernetes cluster. This operator continuously watches the instances (custom resources (CRs)) of this CRD. Any change triggers reconciliation of the Shoot definition to align the Shoot definition with the description provided by the CR.

In addition to aligning the Kubernetes infrastructure, KIM also provides access to these clusters through Kyma Control Plane (KCP) by exposing and rotating their kubeconfigs. Each Kyma runtime has the kubeconfig stored in a Secret on KCP. To address security requirements, KIM also regularly rotates these kubeconfigs.

## Context and Scope

With Kyma Environment Broker (KEB) and Lifecycle Manager, KIM builds the foundation of Kyma runtime.
All backend services run in KCP.

Kyma runs three different control planes, each has a KIM instance deployed:

1. DEV: Used for development and integration of KCP components.

2. STAGE: Runs the current or next release candidates of the KCP components and is used to stabilize these components.
3. PROD: Includes the stabilized components and manages productive Kyma runtimes.

> ## Solution Strategy

Author: Right, makes totally sense

### Sub-Components

To address the requirements for implementing KIM features, the following sub-components have been introduced :

* KIM Snatch

    This mandatory Kyma module belongs to KIM and is automatically installed on all Kyma runtimes. To avoid Kyma workloads from being assigned to worker pools created by customers, KIM Snatch assigns Kyma workloads to the Kyma worker pool.

    For more information, see the [KIM Snatch repository](https://github.com/kyma-project/kim-snatch/tree/main/docs/user).

* Gardener Syncer

    This Kubernetes CronJob synchronizes Seed cluster data from the Gardener cluster to KCP. KEB uses this data for customer input validation (primarily for the "Shoot and Seed Same Region" feature to detect if a Seed cluster exists in a particular Shoot region). The Seed data is stored in a ConfigMap in KCP.

    For more information, see the [Gardener Syncer documentation](https://github.com/kyma-project/gardener-syncer/blob/main/README.md).

## Quality Requirements

Non-technical requirements for KIM are:

> * Performance: KIM processes about 5.000 Runtime instances.
> * Extensibility: To ensure easy extensibility of KIM, the following software patterns are used:  
>     * The state machine pattern for process flow control
>     * The chain of responsibility pattern for rendering the Shoot definitions
> *  Reliability: Since uptime is crucial for KIM, monitoring tools are employed to detect throughput degradation and long-running reconciliation processes, with alerting rules to notify about any SLA violations.
> * Security:  SAP product standards are used during development, and a threat modeling workshop is executed annually.

Reviewer: ```suggestion
* Performance: Processes about 5.000 Runtime instances.
* Extensibility: Ensures easy extensibility by employing the following software patterns:  
    * State machine pattern for process flow control.
    * Chain of responsibility pattern for rendering the Shoot definitions.
*  Reliability: Monitoring tools detect throughput degradation and long-running reconciliation processes, triggering alerts for any SLA violations.
* Security:  Adheres to SAP product standards during development; threat modeling workshops are executed annually.
```

# Kyma Infrastructure Manager

## Introduction

The Kyma Infrastructure Manager (KIM) is responsible for the lifecycle management of SAP Kyma runtimes (SKR). It was replacing the previously used Kyma Provisioner.

KIM is using SAP Gardener for creating and administrating Kubernetes cluster related infrastructure components.

## Goal

KIM is translating Kubernetes cluster related requirements for Kyma runtimes into Gardener's Shoot definitions. It is creating, updating and deleting Shoot definition. While Gardener is reconciling a Kubernetes cluster, KIM is monitoring the progress, reacting on failure cases but also ensuring the cluster is prepared for running the Kyma software.

## Constraints

Kyma customers have multiple options to adjust their Kyma runtime  (e.g. define cluster sizes, configure OIDC authentication providers and administrators, introduce worker pools etc.) KIM has to ensure the created Kubernetes infrastructure from Gardener is aligned with the latest Kyma customer configuration. The alignment has to happen with minimal delay.

> The maintainability of the previous Provisioner was a major drawback and new features required usually modification at multiple places. A simple and easy extendable architecture of KIM has to address these points. The reconciliation flow has to be easy adjust- and extendable.

Author: ```suggestion
```

|Runtime Kubeconfig Custom Resource|RuntimeKubeconfigCR|This CR is used to administrate and rotate the kubeconfig of a Kyma runtime. It is created and managed by KIM.|

### Process Flow

1. The customer requests  a Kyma runtime via BTP. The request is received by KEB which creates/updates an `RuntimeCR` instance. this `RuntimeCR` describes how the infrastructure of the Kubernetes cluster.
2. KIM is detecting the change of this `RuntimeCR` and triggers the reconciliation process.
3. KIM converts the information into a Shoot definition for Gardener.
4. After the cluster was created, KIM creates a `RuntimeKubeconfigCR` which include required data for fetching and rotating the cluster Kubeconfig.
5. It fetches the Kubeconfig from Gardener and stores it in a secret on the KCP cluster.

## Operation

### Installation

As runtime is a Kubernetes cluster expected.

KIM is bundled a containerized application. As component of the Kyma Control Plane, the common delivery process of our SRE team (Argo CD) is supported. Deployment is based on HELM.

> The deployment of KIM is self-contained and can be fully automated (no manual pre- or post-actions are required).

Author: ```suggestion
cluster. The HELM-based deployment process is self-contained and fully automated, requiring no manual pre- or post-deployment actions. The SRE team facilitates deployment through a common delivery process using Argo CD.
```

### Sub-Components
During the implementation of paritcular KIM features were sub-components introduced to address the feature-specifc requirements:

* KIM Snatch

    This is a Kyma module which belongs to KIM and was introduced with the support for multiple worker pools. It's a mandatory module: KLM installs it on all Kyma runtimes automatically, customers cannot avoid this or remove the module. Purpose of KIM Snatch is to assign Kyma workloads to the Kyma worker pool. This is needed to avoid that Kyma workloads will be assigned to worker pools created by customers.

    More information are available in the KIM Snatch repository: https://github.com/kyma-project/kim-snatch/tree/main/docs/user

* Gardener Syncer

    This Kubernetes Cronjob synchronizes Seed cluster data from the Gardener cluster to KCP. This data is used by KEB for input validation of customers (primarily for the "Shoot and Seed in same region" feature to detect if a Seed cluster exists in a particular Shoot region).

    The seed data are stored in a config map in KCP.

    More details includes the documentation of the Gardener Syncer: https://github.com/kyma-project/gardener-syncer/blob/main/README.md

> ### ADR
> Please see the ADR (Architectural Decision Record) of KIM tp get more insights about its [software architecture](../adr/001-provisioning.md).

Reviewer: ```suggestion
### ADR

For more information on KIM's software architecture, see the [Architectural Decision Record (ADR)](../adr/001-provisioning.md).
```

Reviewer: Do users need this? If they do, why not extend the Architecture section to include all the necessary information instead of referring to the ADR?

## Quality Requirements

Non-technical requirements for KIM are:

> * Performance
> 
>     Goal was to be able to process +/- 5.000 Runtime instances with KIM. A load and performance test was implemented which confirms the achievement of this goal.
> 

Reviewer: ```suggestion
* Performance: KIM processes about 5.000 Runtime instances.
```

* Extensibility

    Selected software patterns are used to ensure an easy extensibility of KIM:
    
    *  state machine pattern is used for process flow control
    *  chain of responsibility pattern is used for rendering the Shoot definitions

> *  Reliability
> 
>     Uptime is crucial for KIM. Monitoring is detecting throughput degradation and long-running reconciliation processes. Alerting rule are in place to notify the team about SLA violations.
> 

Reviewer: ```suggestion
*  Reliability: Since uptime is crucial for KIM, monitoring tools are employed to detect throughput degradation and long-running reconciliation processes, with alerting rules to notify about any SLA violations.
```

> * Security
> 
>     SAP product standards are used during development and a threat modeling workshop is executed annually.
> 

Reviewer: ```suggestion
* Security:  SAP product standards are used during development, and a threat modeling workshop is executed annually.
```
